sinaloa/src/ec2_instance.rs

```diff
@@ -0,0 +1,317 @@
+    #[error(display = "EC2: Veracruz Socket Error:{:?}", _0)]
+    VeracruzSocketError(#[error(source)] veracruz_utils::VeracruzSocketError),
+    #[error(display = "EC2: Command non-zero status error:{:?}", _0)]
+    CommandNonZeroStatus(i32),
+    #[error(display = "EC2: Nix error:{:?}", _0)]
+    NixError(nix::Error),
+    #[error(display = "EC2: CLI error")]
+    CLIError,
+    #[error(display = "EC2: Unimplemented")]
+    Unimplemented,
+}
+
+impl EC2Instance {
+    pub fn new() -> Result<Self, EC2Error> {
+        let aws_key_name = env::var("AWS_KEY_NAME").expect("Failed to read AWS_KEY_NAME environment variable.");
+        let aws_subnet = env::var("AWS_SUBNET").expect("Failed to read AWS_SUBNET environment variable.");
+        let aws_region = env::var("AWS_REGION").expect("Failed to read AWS_REGION environment variable.");
+        let aws_security_group_id = env::var("AWS_SECURITY_GROUP_ID").expect("Failed to read AWS_SECURITY_GROUP_ID environment variable.");
+
+        let subnet_option = format!("--subnet-id={:}", aws_subnet);
+
+        let ec2_result = Command::new("/usr/local/bin/aws")
+            .args(&["ec2", "run-instances", "--image-id", "ami-037dd1d3f98da4d50",
+                    "--instance-type", "m5.xlarge", "--enclave-options", "Enabled=true",
+                    "--region",  &aws_region, "--key-name", &aws_key_name,
+                    &subnet_option, "--security-group-ids", &aws_security_group_id,
+                    "--associate-public-ip-address",
+                    "--tag-specifications",  "ResourceType=instance,Tags=[{Key=Veracruz,Value=RootEnclave}]"])
+                        .output()
+            .map_err(|err| {
+                println!("EC2Instance::new failed to start ec2 instance:{:?}", err);
+                EC2Error::IOError(err)
+            })?;
+        if !ec2_result.status.success() {
+            let ec2_result_text = std::str::from_utf8(&ec2_result.stderr)
+                .map_err(|err| EC2Error::Utf8Error(err))?;
+            println!("ec2 result Stdout:{:}", ec2_result_text);
+            return Err(EC2Error::CLIError);
+        }
+
+        let ec2_result_text = std::str::from_utf8(&ec2_result.stdout)
+            .map_err(|err| EC2Error::Utf8Error(err))?;
+        let ec2_data: Value = serde_json::from_str(ec2_result_text)
+            .map_err(|err| EC2Error::SerdeJsonError(err))?;
+
+        let instance_id: &str = match &ec2_data["Instances"][0]["InstanceId"] {
+            Value::String(value) => value,
+            _ => return Err(EC2Error::IncorrectJson),
+        };
+        let private_ip: &str = match &ec2_data["Instances"][0]["PrivateIpAddress"] {
+            Value::String(value) => value,
+            _ => return Err(EC2Error::IncorrectJson),
+        };
+        println!("EC2Instance instance_id:{:?}", instance_id);
+        println!("EC2Instance private_ip:{:?}", private_ip);
+
+        std::thread::sleep(std::time::Duration::from_millis(30000));
```

Comment:
Why the `sleep` here, and why for that duration?

Reply:
I should add a comment, but I will also explain it here.

When you start an EC2 instance, the CLI returns to you immediately, with the details of the instance. However, that instance is not yet completed booting. The networking is not up, yet, and the OS on the instance hasn't completed it's boot sequence. I have found that 30s is a sufficient amount of time to let this happen. Could it be less? In most cases, probably yes, but AWS gives you no promises about the amount of time this will take (as they really can't, because they can't really make promises about how long an arbitrary OS takes to boot).

Reply:
Should it be configurable then? (an environment variable maybe?)

Reply:
I'm not sure if configurability is useful here, as there's really no control over how long it takes. I've found heuristically that 30 seconds works (although this being too low may be the cause of some spurious failures that I'm seeing).
Making it configurable encourages people to change it, and the failure mode isn't obvious.